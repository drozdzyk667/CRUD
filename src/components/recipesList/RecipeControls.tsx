import React from 'react';
import Icon from 'Icons';
import { ButtonsContainer, Button } from '../styled';
import { makeStyles, Box } from '@material-ui/core';
import {
  Recipe,
  Names,
  Colors,
  WARNING_MESSAGE
} from 'pages/recipes/Recipes.constants';
import ConfirmationModal from '../confirmation';

const useStyles = makeStyles({
  container: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center'
  },
  text: {
    margin: '0 0.5em'
  }
});

interface RecipeControlProps {
  recipe: Recipe;
  handleOpenForm: (type: string, element?: Recipe) => void;
  handleDeleteRecipe: (id: number) => void;
}

const RecipeControls: React.FC<RecipeControlProps> = ({
  recipe,
  handleOpenForm,
  handleDeleteRecipe
}) => {
  const classes = useStyles();
  const [isDeleteModal, setIsDeleteModal] = React.useState(false);

  const toggleConfirmationModal = () => {
    setIsDeleteModal(isDeleteModal => !isDeleteModal);
  };

  const handleConfirmDelete = () => {
    handleDeleteRecipe(recipe.id);
    setIsDeleteModal(false);
  };

  return (
    <>
      <ButtonsContainer>
        <Button background={Colors.delete} onClick={toggleConfirmationModal}>
          <Box className={classes.container}>
            <span className={classes.text}>{Names.delete}</span>
            <Icon name='Delete' />
          </Box>
        </Button>
        <Button
          black
          background={Colors.edit}
          onClick={() => handleOpenForm(Names.edit, recipe)}
        >
          <Box className={classes.container}>
            <span className={classes.text}>{Names.edit}</span>
            <Icon name='Edit' />
          </Box>
        </Button>
      </ButtonsContainer>
      <ConfirmationModal
        isDeleteModal={isDeleteModal}
        handleConfirmDelete={handleConfirmDelete}
        toggleConfirmationModal={toggleConfirmationModal}
        message={WARNING_MESSAGE}
      />
    </>
  );
};

export default RecipeControls;
