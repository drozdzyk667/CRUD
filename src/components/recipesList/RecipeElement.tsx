import React from 'react';
import {
  Accordion,
  AccordionSummary,
  AccordionDetails,
  Typography,
  makeStyles,
  Theme,
  Box,
  AccordionActions
} from '@material-ui/core';
import { IngredientContainer } from '../styled';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import { Recipe, Names, Colors } from 'pages/recipes/Recipes.constants';

const useStyles = makeStyles((theme: Theme) => ({
  root: {
    boxShadow: theme.shadows[4],
    paddingBottom: '1em'
  },
  summary: {
    background: Colors.lightGray,
    boxShadow: theme.shadows[2]
  },
  details: {
    display: 'flex',
    flexDirection: 'column'
  },
  title: {
    padding: '1em',
    fontWeight: 'bold',
    margin: 'auto auto',
    marginBottom: '50px',
    borderBottom: `2px solid ${Colors.gray}`
  },
  textWrap: {
    width: 'calc(100vw - 120px)',
    overflow: 'hidden',
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis'
  }
}));

interface RecipeProps {
  recipe: Recipe;
  children: React.ReactNode;
  handleExpandOnlyOneRecipe: (id: number) => void;
}

const RecipeElement: React.FC<RecipeProps> = ({
  recipe,
  children,
  handleExpandOnlyOneRecipe
}) => {
  const classes = useStyles();

  return (
    <>
      <Accordion
        expanded={recipe.isExpanded}
        onChange={() => handleExpandOnlyOneRecipe(recipe.id)}
        className={classes.root}
      >
        <AccordionSummary
          className={classes.summary}
          expandIcon={<ExpandMoreIcon />}
        >
          <Typography className={classes.textWrap} variant='h4'>
            {recipe.recipeName}
          </Typography>
        </AccordionSummary>
        <AccordionDetails className={classes.details}>
          <Typography
            data-testid={Names.ingredients}
            className={classes.title}
            variant='h5'
          >
            {Names.ingredients}
          </Typography>
          <Box>
            {recipe.ingredients.map((ingredient: string, index: number) => (
              <IngredientContainer
                className={classes.textWrap}
                key={`${ingredient}-${index}`}
              >
                {ingredient}
              </IngredientContainer>
            ))}
          </Box>
          <AccordionActions>{children}</AccordionActions>
        </AccordionDetails>
      </Accordion>
    </>
  );
};

export default RecipeElement;
