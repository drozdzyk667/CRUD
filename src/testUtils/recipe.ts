export const RecipeMock = {
  id: 1,
  recipeName: 'Test Recipe',
  isExpanded: false,
  ingredients: ['test1', 'test2', 'test3']
};
