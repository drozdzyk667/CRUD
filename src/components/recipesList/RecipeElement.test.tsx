import React from 'react';
import { shallow, mount } from 'enzyme';
import RecipeElement from './RecipeElement';
import RecipeControls from './RecipeControls';
import { Names } from 'pages/recipes/Recipes.constants';
import { Typography } from '@material-ui/core';
import { RecipeMock } from 'testUtils/recipe';

describe('RecipeElement', () => {
  const clickFn = jest.fn();
  const clickFnOpenForm = jest.fn();
  const clickFnDeleteRecipe = jest.fn();
  const children = (
    <RecipeControls
      handleOpenForm={clickFnOpenForm}
      handleDeleteRecipe={clickFnDeleteRecipe}
      recipe={RecipeMock}
    />
  );
  it('should render correctly ', () => {
    const component = shallow(
      <RecipeElement handleExpandOnlyOneRecipe={clickFn} recipe={RecipeMock}>
        {children}
      </RecipeElement>
    );
    expect(component).toMatchSnapshot();
  });
  it('should render recipe name', () => {
    const component = shallow(
      <RecipeElement handleExpandOnlyOneRecipe={clickFn} recipe={RecipeMock}>
        {children}
      </RecipeElement>
    );
    const name = component.find(Typography);
    expect(name.first().text()).toEqual('Test Recipe');
  });
  it('should render ingredients header', () => {
    const component = shallow(
      <RecipeElement handleExpandOnlyOneRecipe={clickFn} recipe={RecipeMock}>
        {children}
      </RecipeElement>
    );
    const ingredients = component.find({
      'data-testid': Names.ingredients
    });
    expect(ingredients.text()).toEqual('Ingredients');
  });
  it('test if recipe is collapsed', () => {
    const component = mount(
      <RecipeElement handleExpandOnlyOneRecipe={clickFn} recipe={RecipeMock}>
        {children}
      </RecipeElement>
    );
    expect(component.prop('recipe').isExpanded).toEqual(false);
  });

  it('test if recipe render props correctly', () => {
    const component = mount(
      <RecipeElement handleExpandOnlyOneRecipe={clickFn} recipe={RecipeMock}>
        {children}
      </RecipeElement>
    );
    expect(component.prop('recipe')).toEqual({
      id: 1,
      recipeName: 'Test Recipe',
      isExpanded: false,
      ingredients: ['test1', 'test2', 'test3']
    });
  });
});
