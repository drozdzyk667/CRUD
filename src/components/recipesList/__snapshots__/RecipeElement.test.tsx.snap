// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeElement should render correctly  1`] = `
<Fragment>
  <WithStyles(ForwardRef(Accordion))
    className="makeStyles-root-1"
    expanded={false}
    onChange={[Function]}
  >
    <WithStyles(ForwardRef(AccordionSummary))
      className="makeStyles-summary-2"
      expandIcon={<UNDEFINED />}
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-textWrap-5"
        variant="h4"
      >
        Test Recipe
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(AccordionSummary))>
    <WithStyles(ForwardRef(AccordionDetails))
      className="makeStyles-details-3"
    >
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-title-4"
        data-testid="Ingredients"
        variant="h5"
      >
        Ingredients
      </WithStyles(ForwardRef(Typography))>
      <Styled(MuiBox)>
        <styled.div
          className="makeStyles-textWrap-5"
          key="test1-0"
        >
          test1
        </styled.div>
        <styled.div
          className="makeStyles-textWrap-5"
          key="test2-1"
        >
          test2
        </styled.div>
        <styled.div
          className="makeStyles-textWrap-5"
          key="test3-2"
        >
          test3
        </styled.div>
      </Styled(MuiBox)>
      <WithStyles(ForwardRef(AccordionActions))>
        <RecipeControls
          handleDeleteRecipe={[MockFunction]}
          handleOpenForm={[MockFunction]}
          recipe={
            Object {
              "id": 1,
              "ingredients": Array [
                "test1",
                "test2",
                "test3",
              ],
              "isExpanded": false,
              "recipeName": "Test Recipe",
            }
          }
        />
      </WithStyles(ForwardRef(AccordionActions))>
    </WithStyles(ForwardRef(AccordionDetails))>
  </WithStyles(ForwardRef(Accordion))>
</Fragment>
`;
